### ISSUE TITLE
Insufficient C++17 compiler support check in CMake

### ISSUE THREAD

**Issue by davidscn** [precice/precice#1617]
**Please describe the problem you are trying to solve.**

Using GCC 9.2.0 and our latest develop, I am able to configure precice using `CMake`. However, at compile time, the compiler generates the following error message:

```bash
~/precice/src/com/SerializedPartitioning.cpp: In static member function ‘static precice::com::serialize::SerializedConnectionMap precice::com::serialize::SerializedConnectionMap::serialize(const ConnectionMap&)’:
~/precice/src/com/SerializedPartitioning.cpp:26:8: error: no matching function for call to ‘transform_reduce(std::map<int, std::vector<int> >::const_iterator, std::map<int, std::vector<int> >::const_iterator, int, std::plus<long unsigned int>, precice::com::serialize::SerializedConnectionMap::serialize(const ConnectionMap&)::<lambda(const auto:1&)>)’
   26 |       });
      |        ^
```

The problem here is that this GCC version does not fully support the required [`std::transform_reduce`](https://en.cppreference.com/w/cpp/algorithm/transform_reduce) C++17 feature.

**Describe the solution you propose.**

We should catch this already at compile time and inform the user about this. Looking at our CMake C++ checks, our standard support check [only checks for C++11 compliance](https://github.com/precice/precice/blob/develop/cmake/CheckCXX11Library.cmake). It would be easy to either extend the check to C++17 or add an additional C++17 check.

For me, it was sufficient to modify the C++11 check for 

```cpp
#include <execution>
```

**Describe alternatives you've considered**
Do nothing and let the user figure out.

**Comment by MakisH** [comment]
Another alternative: do we really need to use `std::transform_reduce`? Could we replace it with a combination of older/other algorithms?

GCC 9.3.0 (I guess that's the one you mean, right?) is the default on Ubuntu 20.04, so we should support it for another year, based on our policy: https://precice.org/installation-packages.html#ubuntu

**Comment by davidscn** [comment]
> Another alternative: do we really need to use std::transform_reduce? Could we replace it with a combination of older/other algorithms?

Note that we use it in several places and I just put here the very first one. Since we are using C++17, I would not try to backport now our implementation.

> GCC 9.3.0 (I guess that's the one you mean, right?) is the default on Ubuntu 20.04

No, I mean GCC 9.2. Ubuntu 20.04 is covered in our CI.

**Comment by fsimonis** [comment]
A big open question in need of investigation is why the macOS pipeline doesn't fail, given the default standard library used by macOS doesn't contain `transform_reduce`.

**Comment by fsimonis** [comment]
I checked the sources of libc++ and `std::transform_reduce` was added with llvm 5.0.0 .
GCC libstdc++, added versions without execution policies in version 9.3.

**Conclusion:** We can use `std::transform_reduce`. The baseline gcc version 9.4 shipped with Ubuntu 20.04 is a hard requirement.

**Comment by davidscn** [comment]
> Conclusion: We can use std::transform_reduce. The baseline gcc version 9.4 shipped with Ubuntu 20.04 is a hard requirement.

IIRC, installing `tbb` along with lower gcc versions could also work (or was it, what later versions relied on?). Point of this issue is also to forward lacking support of features at configuration time rather than compile time (as you know).

**Comment by fsimonis** [comment]
Not sure about tbb, but I definitely agree that we should add small checks for such cases.

I'll have a look and document this for future cases.
